--- BSB_lan.ino.orig    2020-01-29 17:29:51.963830790 +0300
+++ BSB_lan.ino 2020-01-30 17:14:59.671539315 +0300
@@ -462,7 +462,7 @@
 // char _ipstr[20];    // addr in format xxx.yyy.zzz.aaa
 // byte __remoteIP[4] = {0,0,0,0};   // IP address in bin format

-#ifdef LOGGER
+#if defined LOGGER || defined WEBSERVER
 #if defined(__SAM3X8E__)
   #include <SD.h>
 #else
@@ -4157,6 +4157,62 @@
   return log_parameter;
 }

+/** *****************************************************************
+ *  Function: transmitFile
+ *  Does: transmit file from SD card to network client
+ *  Pass parameters:
+ *  File dataFile - opened file
+ * Parameters passed back:
+ *   none
+ * Function value returned:
+ *   none
+ * Global resources used:
+ *   client instance
+ * *************************************************************** */
+void transmitFile(File dataFile) {
+  int logbuflen = 512;
+  byte loglineBuf[logbuflen];
+  int chars_read = dataFile.read(&loglineBuf , logbuflen);
+  while (chars_read == logbuflen) {
+    client.write(loglineBuf, logbuflen);
+    chars_read = dataFile.read(&loglineBuf , logbuflen);
+    }
+  if (chars_read > 0) client.write(loglineBuf, chars_read);
+}
+
+/** *****************************************************************
+ *  Function: bufferedprint and bufferedprintln
+ *  Does: do buffered print to network client. Increasing net perfomance 2~50 times
+ *  Pass parameters:
+ *  WiFiEspClient/EthernetClient &cl
+ *  PGM_P outstr
+ * Parameters passed back:
+ *   none
+ * Function value returned:
+ *   none
+ * Global resources used:
+ *   buffer variable
+ * *************************************************************** */
+
+#ifdef WIFI
+void bufferedprint(WiFiEspClient& cl, PGM_P outstr){
+#else
+void bufferedprint(EthernetClient& cl, PGM_P outstr){
+#endif
+  strncpy_P(buffer, outstr, BUFLEN);
+  buffer[BUFLEN - 1] = 0;
+  cl.print(buffer);
+}
+#ifdef WIFI
+void bufferedprintln(WiFiEspClient& cl, PGM_P outstr){
+#else
+void bufferedprintln(EthernetClient& cl, PGM_P outstr){
+#endif
+  strncpy_P(buffer, outstr, BUFLEN - 2);
+  strcat_P(buffer, PSTR("\n"));
+  buffer[BUFLEN - 1] = 0;
+  cl.print(buffer);
+}

 /** *****************************************************************
  *  Function:
@@ -4832,29 +4888,45 @@
         }
         // Got an EOL character
         DebugOutput.println();
-        // perform HTTP-Authentification by reading the remaining client data and look for credentials
-#ifdef USER_PASS_B64

-        char linebuf[80];
+        // perform HTTP-Authentification by reading the remaining client data and look for credentials
+        // Parsing headers
+        //char linebuf[80]; i have 'buffer' variable. I will use 'buffer' instead 'linebuf'
         uint8_t charcount=0;
-        boolean authenticated=false;
-        memset(linebuf,0,sizeof(linebuf));
+        uint8_t httpflags = 0; //bit 0 - authenticated: 0 - no, 1 - yes
+                               //bit 1 - client browser accept gzip: 0 - no, 2 - yes
+                               //bit 2 - HEAD request received from client: 0 - no, 4 - yes
+                               //...
+                               //bit 7 - send HTML fragment only, without header and footer. For external webserver. 0 - full HTML, 128 - fragment
+        memset(buffer,0,sizeof(buffer));
         boolean currentLineIsBlank = false;
         while (client.connected()) {
           if (client.available()) {
             char c = client.read();
-            linebuf[charcount]=c;
-            if (charcount<sizeof(linebuf)-1) charcount++;
+            buffer[charcount]=c;
+            if (charcount<sizeof(buffer)-1) charcount++;
             if (c == '\n' && currentLineIsBlank) {
               break;
             }
             if (c == '\n') {
               // you're starting a new line
               currentLineIsBlank = true;
-              if (strstr(linebuf,"Authorization: Basic")!=0 && strstr(linebuf,USER_PASS_B64)!=0) {
-                authenticated=true;
+#ifdef WEBSERVER
+              //Execute only if flag not set because strstr more expensive than bitwise operation
+              if (!(httpflags & 2) && strstr_P(buffer,PSTR("Accept-Encoding")) != 0 && strstr_P(buffer+16, PSTR("gzip")) != 0) {
+                httpflags |= 2;
               }
-              memset(linebuf,0,sizeof(linebuf));
+#ifdef USER_PASS_B64
+             else
+#endif
+#endif
+#ifdef USER_PASS_B64
+              //Execute only if flag not set because strstr more expensive than bitwise operation
+              if (!(httpflags & 1) && strstr_P(buffer,PSTR("Authorization: Basic"))!=0 && strstr(buffer,USER_PASS_B64)!=0) {
+                httpflags |= 1;
+              }
+#endif
+              memset(buffer,0, charcount);
               charcount=0;
             } else if (c != '\r') {
               // you've gotten a character on the current line
@@ -4862,8 +4934,9 @@
             }
           }
         }
+#ifdef USER_PASS_B64
         // if no credentials found in HTTP header, send 401 Authorization Required
-        if (!authenticated) {
+        if (!(httpflags & 1)) {
           client.println(F("HTTP/1.1 401 Authorization Required"));
           client.println(F("WWW-Authenticate: Basic realm=\"Secure Area\""));
           client.println(F("Content-Type: text/html"));
@@ -4876,12 +4949,15 @@
         }
         // otherwise continue like normal
 #endif
-
         // Flush any remaining bytes from the client buffer
 //        client.flush();
         // GET / HTTP/1.1 (anforderung website)
         // GET /710 HTTP/1.0 (befehlseingabe)
         String urlString = String(cLineBuffer);
+#ifdef WEBSERVER
+        // Check for HEAD request (for file caching)
+        if (urlString.substring(0, urlString.indexOf('/')).indexOf("HEAD") != -1 ) httpflags |= 4;
+#endif
         urlString = urlString.substring(urlString.indexOf('/'), urlString.indexOf(' ', urlString.indexOf('/')));
         DebugOutput.println(urlString);
         urlString.toCharArray(cLineBuffer, MaxArrayElement);
@@ -4897,14 +4973,22 @@
 // IPWE END

         if (urlString == "/favicon.ico") {
-          client.println(F("HTTP/1.1 200 OK"));
-          client.println(F("Content-Type: image/x-icon"));
-          client.println();
+//          strcpy_P(buffer, PSTR("HTTP/1.1 200 OK\nContent-Type: image/x-icon\n\n"));
+//          client.print(buffer);
+          bufferedprint(client, PSTR("HTTP/1.1 200 OK\nContent-Type: image/x-icon\n\n"));
+#ifdef WEBSERVER
+          File dataFile = SD.open(urlString + 1);
+          if (dataFile) {
+            transmitFile(dataFile);
+            dataFile.close();
+            } else {
 #if defined(__SAM3X8E__)
           printPStr(favicon, sizeof(favicon));
 #else
           printPStr(pgm_get_far_address(favicon), sizeof(favicon));
 #endif
+            }
+#endif
           break;
         }

@@ -4926,21 +5010,157 @@
         }
         *p='/';
 #endif
+
+#ifdef WEBSERVER
+        if(!strcmp(p,"/")){
+          urlString = F("index.html");
+        }
+        else
+          urlString = String(p + 1);
+        DebugOutput.println("URL: " + urlString);
+        int mimetype = 0; //unknown MIME type
+        if (urlString.endsWith(F(".html")) || urlString.endsWith(F(".htm"))) mimetype = 1;
+        else if(urlString.endsWith(F(".css"))) mimetype = 2;
+        else if(urlString.endsWith(F(".js"))) mimetype = 3;
+        else if(urlString.endsWith(F(".xml"))) mimetype = 4;
+        else if(urlString.endsWith(F(".txt"))) mimetype = 5;
+        else if(urlString.endsWith(F(".jpg"))) mimetype = 101;
+        else if(urlString.endsWith(F(".gif"))) mimetype = 102;
+        else if(urlString.endsWith(F(".svg"))) mimetype = 103;
+        else if(urlString.endsWith(F(".png"))) mimetype = 104;
+        else if(urlString.endsWith(F(".ico"))) mimetype = 105;
+        else if(urlString.endsWith(F(".gz"))) mimetype = 201;
+        // You can add more MIME types here
+
+        if(mimetype)  {
+          File dataFile;
+          const char *getfarstrings;
+
+          // client browser accept gzip
+          if ((httpflags & 2)) dataFile = SD.open(urlString + ".gz");
+          if (!dataFile) {
+            // reuse httpflags
+            bitClear(httpflags, 1); //can't use gzip because no gzipped file
+            dataFile = SD.open(urlString);
+          }
+          // if the file is available, read from it:
+          if (dataFile) {
+            DebugOutput.print(F("File opened from SD: ")); DebugOutput.println(urlString);
+            strcpy_P(buffer, PSTR("HTTP/1.1 200 OK\nContent-Type: "));
+            switch(mimetype){
+              case 1: getfarstrings = PSTR("text/html"); break;
+              case 2: getfarstrings = PSTR("text/css"); break;
+              case 3: getfarstrings = PSTR("application/x-javascript"); break;
+              case 4: getfarstrings = PSTR("application/xml"); break;
+              // case 5 below
+              case 101: getfarstrings = PSTR("image/jpeg"); break;
+              case 102: getfarstrings = PSTR("image/gif"); break;
+              case 103: getfarstrings = PSTR("image/svg"); break;
+              case 104: getfarstrings = PSTR("image/png"); break;
+              case 105: getfarstrings = PSTR("image/x-icon"); break;
+              case 201: getfarstrings = PSTR("application/x-gzip"); break;
+              case 5:
+              default: getfarstrings = PSTR("text");
+            }
+            strcat_P(buffer, getfarstrings);
+            strcpy_P(outBuf, PSTR("\nContent-Length: %d\n"));
+            sprintf(buffer + strlen(buffer), outBuf, dataFile.size());
+            if((httpflags & 2)) strcat_P(buffer, PSTR("Content-Encoding: gzip\n"));
+            dir_t d;
+            if (dataFile.dirEntry(&d)) {
+              char monthname[4];
+              char downame[4];
+              uint16_t lastWrtYr =  (FAT_YEAR(d.lastWriteDate));
+              byte monthval = FAT_MONTH(d.lastWriteDate);
+              byte dayval = FAT_DAY(d.lastWriteDate);
+              switch (dayofweek((uint8_t)dayval, (uint8_t)monthval, lastWrtYr))
+              {
+                case 1: getfarstrings = PSTR("Mon"); break;
+                case 2: getfarstrings = PSTR("Tue"); break;
+                case 3: getfarstrings = PSTR("Wed"); break;
+                case 4: getfarstrings = PSTR("Thu"); break;
+                case 5: getfarstrings = PSTR("Fri"); break;
+                case 6: getfarstrings = PSTR("Sat"); break;
+                case 7: getfarstrings = PSTR("Sun"); break;
+                default: getfarstrings = PSTR("ERR"); break;
+              }
+              strcpy_P(downame, getfarstrings);
+
+              switch (monthval)
+              {
+                case 1: getfarstrings = PSTR("Jan"); break;
+                case 2: getfarstrings = PSTR("Feb"); break;
+                case 3: getfarstrings = PSTR("Mar"); break;
+                case 4: getfarstrings = PSTR("Apr"); break;
+                case 5: getfarstrings = PSTR("May"); break;
+                case 6: getfarstrings = PSTR("Jun"); break;
+                case 7: getfarstrings = PSTR("Jul"); break;
+                case 8: getfarstrings = PSTR("Aug"); break;
+                case 9: getfarstrings = PSTR("Sep"); break;
+                case 10: getfarstrings = PSTR("Oct"); break;
+                case 11: getfarstrings = PSTR("Nov"); break;
+                case 12: getfarstrings = PSTR("Dec"); break;
+                default: getfarstrings = PSTR("ERR"); break;
+              }
+              strcpy_P(monthname, getfarstrings);
+              strcpy_P(outBuf, PSTR("Last-Modified: %s, %02d %s %d %02d:%02d:%02d GMT\n"));
+              sprintf(buffer + strlen(buffer), outBuf, downame, dayval, monthname, lastWrtYr, FAT_HOUR(d.lastWriteTime), FAT_MINUTE(d.lastWriteTime), FAT_SECOND(d.lastWriteTime));
+            }
+            strcat_P(buffer, PSTR("Cache-Control: max-age=2592000, private\n\n")); //max-age=84400 = one day, max-age=2592000 = 30 days. Last string in header, double \n
+            client.print(buffer);
+
+            //HEAD request received
+            if (!(httpflags & 4)) {
+              transmitFile(dataFile);
+            }
+            strcpy_P(buffer, ((httpflags & 4)?PSTR("HEAD"):PSTR("GET"))); strcat_P(buffer, PSTR(" request received\n"));
+            DebugOutput.print(buffer);
+
+            dataFile.close();
+          }
+          else
+          {
+          // simply print the website if no index.html on SD card
+            if(!strcmp(p,"/")){
+              webPrintSite();
+              break;
+            }
+            strcpy_P(buffer, PSTR("HTTP/1.1 404 Not Found\nContent-Type: text/html\n\n<h2>File not found!</h2><br>File name: "));
+            uint16_t x = strlen(buffer);
+            urlString.toCharArray(buffer + x, BUFLEN - x);
+            client.print(buffer);
+           }
+          client.flush();
+          break;
+        }
+#endif
+
         if (p[1] != 'J') {
           client.flush();
         }
+#ifndef WEBSERVER
         // simply print the website
         if(!strcmp(p,"/")){
           webPrintSite();
           break;
         }
+#endif

         // Answer to unknown requests
-        if(!isdigit(p[1]) && strchr("ABCDEGHIJKLMNOPQRSTUVXY",p[1])==NULL){
+        if(!isdigit(p[1]) && strchr("ABCDEGHIJKLMNOPQRSTUVWXY",p[1])==NULL){
           webPrintHeader();
           webPrintFooter();
           break;
         }
+
+#ifdef WEBSERVER
+        //Send HTML pages without header and footer (For external interface)
+        if(p[1]=='W'){
+          p++;
+          httpflags |= 128;
+        }
+#endif
+
         // setting verbosity level
         if(p[1]=='V'){
           p+=2;
@@ -5177,7 +5397,7 @@
         }

         if(p[1]=='Q') {
-          webPrintHeader();
+          if(!(httpflags & 128)) webPrintHeader();

           client.print(F(MENU_TEXT_VER ": "));
           client.print(BSB_VERSION);
@@ -5358,7 +5578,7 @@

           client.println(F("<BR>" MENU_TEXT_QFE ".<BR>"));
           bus.setBusType(bus.getBusType(), myAddr, destAddr);   // return to original destination address
-          webPrintFooter();
+          if(!(httpflags & 128)) webPrintFooter();
           break;
         }

@@ -5723,6 +5943,7 @@
               client.println(F(MENU_TEXT_DTF));
             }
             webPrintFooter();
+#if !defined WEBSERVER && !defined USEEXTERNALINTERFACE
           } else if (p[2]=='G') {
             webPrintHeader();
             client.println(F("<A HREF='D'>" MENU_TEXT_DTD "</A><div align=center></div>"));
@@ -5742,17 +5963,7 @@
             if (dataFile) {

               unsigned long startdump = millis();
-              int logbuflen = 512;
-              byte loglineBuf[logbuflen];
-              int chars_read = 0;
-
-              chars_read = dataFile.read(&loglineBuf , logbuflen);
-              while (chars_read == logbuflen) {
-                client.write(loglineBuf,logbuflen);
-                chars_read = dataFile.read(&loglineBuf , logbuflen);
-              }
-              //final packet
-              if (chars_read > 0) client.write(loglineBuf, chars_read);
+              transmitFile(dataFile);
               dataFile.close();

               DebugOutput.print(F("Duration: "));
@@ -5760,12 +5971,13 @@
             } else {
               client.println(F(MENU_TEXT_DTO));
             }
+#endif
           }
           break;
         }
 #endif
         if (p[1]=='C'){ // dump configuration
-          webPrintHeader();
+          if(!(httpflags & 128)) webPrintHeader();
           client.println(F(MENU_TEXT_CFG "<BR><BR>"));
 //          client.println(F("BSB pins: "));
 //          client.println(bus);
@@ -5930,7 +6142,7 @@
           #endif

           client.println(F("<BR>"));
-          webPrintFooter();
+          if(!(httpflags & 128)) webPrintFooter();

 #if defined(__AVR__)
           DebugOutput.println(F("EEPROM dump:"));
@@ -7022,7 +7234,7 @@
     }
   }

-#ifdef LOGGER
+#if defined LOGGER || defined WEBSERVER
   // disable w5100 while setting up SD
   pinMode(10,OUTPUT);
   digitalWrite(10,HIGH);
